protocol/river-layout-v3.xml

```diff
@@ -0,0 +1,178 @@
+      <description summary="the compositor requires a layout">
+        The compositor sends this event to inform the client that it requires a
+        layout for a set of views.
+
+        The usable width and height height indicate the space in which the
+        client can safely position views without interfering with desktop
+        widgets such as panels.
+
+        The serial of this event is used to identify subsequent requests as
+        belonging to this layout demand. Beware that the client might need
+        to handle multiple layout demands at the same time.
+
+        The server will ignore responses to all but the most recent layout
+        demand. Thus, clients are only required to respond to the most recent
+        layout_demand received. If a newer layout_demand is received before
+        the client has finished responding to an old demand, the client should
+        abort work on the old demand as any further work would be wasted.
+      </description>
+      <arg name="view_count" type="uint" summary="number of views in the layout"/>
+      <arg name="usable_width" type="uint" summary="width of the usable area"/>
+      <arg name="usable_height" type="uint" summary="height of the usable area"/>
+      <arg name="tags" type="uint" summary="tags of the output, 32-bit bitfield"/>
+      <arg name="serial" type="uint" summary="serial of the layout demand"/>
+    </event>
+
+    <request name="push_view_dimensions">
+      <description summary="propose dimensions of the next view">
+        This request proposes a size and position for a view in the layout demand
+        with matching serial.
+
+        A client must send this request for every view that is part of the
+        layout demand. The number of views in the layout is given by the
+        view_count argument of the layout_demand event. Pushing too many or
+        too few view dimensions is a protocol error.
+
+        The x and y coordinates are relative to the usable area of the output,
+        with (0,0) as the top left corner.
+      </description>
+      <arg name="serial" type="uint" summary="serial of layout demand"/>
+      <arg name="x" type="int" summary="x coordinate of view"/>
+      <arg name="y" type="int" summary="y coordinate of view"/>
+      <arg name="width" type="uint" summary="width of view"/>
+      <arg name="height" type="uint" summary="height of view"/>
+    </request>
+
+    <request name="commit">
+      <description summary="commit a layout">
+        This request indicates that the client is done pushing dimensions
+        and the compositor may apply the layout. This completes the layout
+        demand with matching serial, any other requests sent with the serial
+        are a protocol error.
+
+        The compositor is free to use this proposed layout however it chooses,
+        including ignoring it.
+      </description>
+      <arg name="serial" type="uint" summary="serial of layout demand"/>
```

Comment:
I wonder if we should bring back layout names.

As a reminder, back in the earliest drafts we had a parameter here that was basically an arbitrary string passed back to river, with the intention that river would send it to status tracking clients. The idea was to allow status bars to display layout information, for example simple layout names like `[]=` in dwm, or more complex layout stats, like for example how many views are in a monocle layout.

That is certainly something I see people wanting to have.

Reply:
Hmm, this could also be added later in a non-breaking way as an optional request sent before the commit. Having it as part of the commit request would be nicer though.

Do you remember why we didn't end up including layout names in the protocol? I don't and #112 has over 100 comments.

Reply:
I removed it because river-options could be used for this instead.

Reply:
Ah yeah that makes sense. I think it's a very reasonable feature to have so I'll add it back in this PR and have a look at actually exposing it over river-status

Reply:
done :)